apm-sniffer/apm-sdk-plugin/asynchttpclient-2.x-plugin/src/main/java/org/apache/skywalking/apm/plugin/asynchttpclient/v1/AsyncHandlerWrapper.java

```diff
@@ -0,0 +1,148 @@
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.skywalking.apm.plugin.asynchttpclient.v1;
+
+import io.netty.channel.Channel;
+import io.netty.handler.codec.http.HttpHeaders;
+import java.net.InetSocketAddress;
+import java.util.List;
+import javax.net.ssl.SSLSession;
+import org.apache.skywalking.apm.agent.core.context.trace.AbstractSpan;
+import org.asynchttpclient.AsyncCompletionHandlerBase;
+import org.asynchttpclient.AsyncHandler;
+import org.asynchttpclient.HttpResponseBodyPart;
+import org.asynchttpclient.HttpResponseStatus;
+import org.asynchttpclient.netty.request.NettyRequest;
+
+/**
+ * {@link AsyncHandlerWrapper} wrapper the {@link AsyncHandler} object for tracing.
+ * if userAsyncHandler is null, we will set {@link AsyncCompletionHandlerBase} to avoid NPE.
+ */
+public class AsyncHandlerWrapper implements AsyncHandler {
+
+    private final AsyncHandler userAsyncHandler;
+    private final AbstractSpan asyncSpan;
+
+    public AsyncHandlerWrapper(AsyncHandler asyncHandler, AbstractSpan span) {
+        this.userAsyncHandler = asyncHandler == null ? new AsyncCompletionHandlerBase() : asyncHandler;
+        this.asyncSpan = span;
+    }
+
+    @Override
+    public State onStatusReceived(final HttpResponseStatus httpResponseStatus) throws Exception {
+        return userAsyncHandler.onStatusReceived(httpResponseStatus);
+    }
+
+    @Override
+    public State onHeadersReceived(final HttpHeaders httpHeaders) throws Exception {
+        return userAsyncHandler.onHeadersReceived(httpHeaders);
+    }
+
+    @Override
+    public State onBodyPartReceived(final HttpResponseBodyPart httpResponseBodyPart) throws Exception {
+        return userAsyncHandler.onBodyPartReceived(httpResponseBodyPart);
+    }
+
+    @Override
+    public State onTrailingHeadersReceived(final HttpHeaders headers) throws Exception {
+        return userAsyncHandler.onTrailingHeadersReceived(headers);
+    }
+
+    @Override
+    public void onThrowable(final Throwable throwable) {
+        asyncSpan.log(throwable);
+        asyncSpan.asyncFinish();
+        userAsyncHandler.onThrowable(throwable);
+    }
+
+    @Override
+    public Object onCompleted() throws Exception {
+        asyncSpan.asyncFinish();
```

Comment:
This method should be `try-catch`. This could trigger the exception if the plugin has bug, or the target lib has bug.

Reply:
done.